stl/inc/syncstream

```diff
@@ -0,0 +1,312 @@
+// syncstream standard header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef _SYNCSTREAM_
+#define _SYNCSTREAM_
+#include <yvals.h>
+#if _STL_COMPILER_PREPROCESSOR
+
+#ifdef _M_CEE_PURE
+#error <syncstream> is not supported when compiling with /clr:pure
+#endif // _M_CEE_PURE
+
+#ifndef __cpp_lib_syncbuf
+#pragma message("The contents of <syncstream> are available only with C++20 syncbuf support.")
+#else // ^^^ !defined(__cpp_lib_syncbuf) / defined(__cpp_lib_syncbuf) vvv
+#include <map>
+#include <memory>
+#include <shared_mutex>
+#include <streambuf>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#pragma push_macro("emit")
+#undef new
+#undef emit
+
+_STD_BEGIN
+
+struct _Mutex_count_pair {
+    shared_mutex _Mutex;
+    uint64_t _Ref_count = 0;
+};
+
+_CRTIMP2_PURE map<void*, _Mutex_count_pair>& __cdecl _Get_mutex_map();
+_CRTIMP2_PURE shared_mutex& __cdecl _Get_map_mutex();
+
+shared_mutex& _Get_mutex_for_instance(void* _Ptr) {
+    scoped_lock _Guard(_Get_map_mutex());
+    auto& _Mutex_map         = _Get_mutex_map();
+    auto _Instance_mutex_itr = _Mutex_map.find(_Ptr);
+    _STL_ASSERT(_Instance_mutex_itr != _Mutex_map.end(), "No mutex exists for given instance!");
+    return _Instance_mutex_itr->second._Mutex;
+}
+void _Add_mutex_for_instance_or_increment(void* _Ptr) noexcept {
+    scoped_lock _Guard(_Get_map_mutex());
+    auto& _Mutex_map = _Get_mutex_map();
+    _Mutex_map.try_emplace(_Ptr).first->second._Ref_count++;
+}
+void _Delete_mutex_for_instance_or_decrement(void* _Ptr) noexcept {
+    scoped_lock _Guard(_Get_map_mutex());
+    auto& _Mutex_map         = _Get_mutex_map();
+    auto _Instance_mutex_itr = _Mutex_map.find(_Ptr);
+    if (_Instance_mutex_itr == _Mutex_map.end())
+        return;
+    _Instance_mutex_itr->second._Ref_count--;
+    if (_Instance_mutex_itr->second._Ref_count == 0) {
+        _Mutex_map.erase(_Ptr);
+    }
+}
+
+// CLASS TEMPLATE basic_syncbuf
+template <class _Elem, class _Traits, class _Alloc>
+class basic_syncbuf : public basic_streambuf<_Elem, _Traits> {
+public:
+    using int_type       = typename _Traits::int_type;
+    using pos_type       = typename _Traits::pos_type;
+    using off_type       = typename _Traits::off_type;
+    using allocator_type = _Alloc;
+    using streambuf_type = basic_streambuf<_Elem, _Traits>;
+
+    using pointer   = typename allocator_traits<_Alloc>::pointer;
+    using size_type = typename allocator_traits<_Alloc>::size_type;
+
+    basic_syncbuf() : basic_syncbuf(nullptr){};
+
+    explicit basic_syncbuf(streambuf_type* _StrBuf) : basic_syncbuf(_StrBuf, _Alloc{}) {}
+
+    basic_syncbuf(streambuf_type* _StrBuf, const _Alloc& _Al) : _Wrapped(_StrBuf), _Al(_Al) {

```

Comment:
We conventionally avoid all shadowing, including constructor parameter/data member shadowing. Here, we conventionally add a trailing underscore to the constructor parameter, taking `_Al_`.